primitives/header-chain/src/lib.rs

```diff
@@ -42,3 +42,22 @@ pub trait BaseHeaderChain {
 	/// Returns Some(transaction) if proof is valid and None otherwise.
 	fn verify_transaction_inclusion_proof(proof: &Self::TransactionInclusionProof) -> Option<Self::Transaction>;
 }
+
+/// A trait for verifying whether a header is valid for a particular blockchain.
+pub trait HeaderVerifier {
+	type Header: Parameter;
+	type Extra: Parameter;
+
+	/// Check that a standalone header is well-formed. This does not need to provide any sort
+	/// of ancestry related verification.
+	fn validate_header(header: &Self::Header, extra_data: &Option<Self::Extra>) -> bool;
+}
+
+/// A trait for verifying that a given header has been finalized.
+pub trait FinalityVerifier {
+	type Header: Parameter;
+	type Proof: Parameter;
+
+	/// Verify that the given header has been finalized and is part of the canonical chain.
+	fn verify_finality(header: &Self::Header, proof: &Self::Proof) -> bool;
```

Comment:
Any non-trivial finality verification will require storage access to:
1. Proof ancestry (we build on the previously finalized chain)
2. Check current validator set.

I've made some small proposal how we could handle that in the issue (just expanded it):
```rust
// contains instructions to the upper module what should happen with the headers.
struct ImportOutcome<HeaderId> {
 finalize: Vec<HeaderId>,
 enact: Vec<HeaderId>,
 retract: Vec<HeaderId>,
}

/// T type will give us access to underlying header storage (might be a different trait)
trait ChainVerifier<T: HeaderChain> {
/// Header type
type Header;
/// Extra data for header import (may contain finality proof, receitps, etc)
type HeaderImportExtraData;
/// A way to access custom storage items that are needed by this particular implementation
type CustomStorage;
/// Supported proof kinds
type Proof;
/// A proof error type
type ProofError: core::fmt::Debug;

/// basic (cheap) verification of the header
fn verify_header(Self::Header, Option<Self::HeaderImportExtraData>, &Self::CustomStorage) -> bool;
/// import header against latest state - this should verify finality internally already.
fn import_header(Self::Header, Option<Self::HeaderImportExtraData> &mut Self::CustomStorage) -> ImportOutcome;
/// Check proof correctness (custom storage mutable to allow for caches)
fn verify_proof(Self::Proof, &mut Self::CustomStorage) -> Result<(), Self::ProofError>;
}


/// And the implementation could look like this:
mod impl {
  trait SubstrateExtraStorage<C: ChainVerifier> {
    fn insert_authority_set(&mut self, ...);
  }

  enum Proof {
    TransactionProof(Vec<u8>),
  }

  struct SubstrateLikeChain;
  impl<T: HeaderChain> ChainVerifier<T> for SubstrateLikeChain {
    type Header = sp_core::generic::Header<...>;
    type HeaderImportExtraData = GrandpaFinalityProof;
    type CustomStorage = Box<dyn SubstrateExtraStorage<Self>>;
    type Proof = Proof;
    type ProofError = String;
  
    fn verify_header(h: Self::Header, Option<Self::HeaderImportExtraData>, &Self::CustomStorage) -> bool {
      // Check that it's latest (although that could done externally in the generic bridge pallet, but it's header to show how to access the storage)
      if T::last_finalized_number() > h.number() {
         false
      } else {
         true
      }
    }
    fn import_header(Self::Header, Option<Self::HeaderImportExtraData>, s: &mut Self::CustomStorage) -> ImportOutcome {
         // TODO: verify header, finality proof (optional), check the authority set, enact the authority set if it changes,
         /// mark some more blocks as finalized, etc.
         s.insert_authority_set(...);
         ImportOutcome {
            ...
         }
    }
    fn verify_proof(proof: Self::Proof, p: &mut Self::CustomStorage) -> Result<(), Self::ProofError> {
          let { header_hash, transaction_proof } = extract_tx_proof(proof)?;
          let header = T::header_by_hash(header_hash);
          if verify_merkle_proof(header.extrinsics_root, transaction_proof) { 
             Ok(())
          } else {
             Err("Invalid Proof")
          }
    }
  }
}


```
I don't think that splitting `HeaderVerifier` and `FinalityVerifier` is reasonable, especially in it's current form, since we do both of them anyway.